SmartDeviceLink/SDLCacheFileManager.m

```diff
@@ -0,0 +1,284 @@
+
+typedef NS_ENUM(NSInteger, SDLCacheManagerError) {
+    SDLCacheManagerErrorUpdateArchiveFileFailure = 0
+};
+
+@interface SDLCacheFileManager()
+
+@property (nonatomic, strong) NSURLSession *urlSession;
+@property (strong, nonatomic, readonly, nullable) NSString *cacheFileBaseDirectory;
+@property (strong, nonatomic, readonly, nullable) NSString *archiveFileDirectory;
+@property (weak, nonatomic, nullable) NSURLSessionDataTask *dataTask;
+@property (strong, nonatomic) NSFileManager *fileManager;
+
+@end
+
+@implementation SDLCacheFileManager
+
+- (instancetype)init {
+    self = [super init];
+    if (!self) { return nil; }
+    
+    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
+    configuration.timeoutIntervalForRequest = DefaultConnectionTimeout;
+    configuration.timeoutIntervalForResource = DefaultConnectionTimeout;
+    configuration.requestCachePolicy = NSURLRequestUseProtocolCachePolicy;
+    
+    _urlSession = [NSURLSession sessionWithConfiguration:configuration];
+    _fileManager = [NSFileManager defaultManager];
+
+    return self;
+}
+
+- (void)retrieveImageForRequest:(SDLOnSystemRequest *)request withCompletionHandler:(ImageRetrievalCompletionHandler)completion {
+    NSError *error = nil;
+    UIImage *icon = nil;
+
+    // Check if the directory already exists; if it does not, create it
+    if (![self.fileManager fileExistsAtPath:self.cacheFileBaseDirectory]) {
+        // Create the directory including any intermediate directories if necessary
+        [self.fileManager createDirectoryAtPath:self.cacheFileBaseDirectory withIntermediateDirectories:YES attributes:nil error:&error];
+        
+        if (error != nil) {
+            SDLLogE(@"Could not create directory at specified path: %@; we will attempt to download and return image anyway", error);
+        }
+    }
+    
+    // Attempt to retrieve archive file from directory cache
+    SDLIconArchiveFile *iconArchiveFile = [NSKeyedUnarchiver unarchiveObjectWithFile:self.archiveFileDirectory];
+
+    // Loop through the icons in the cache file and check if its the file we're looking for
+    for (SDLLockScreenIconCache *iconCache in iconArchiveFile.lockScreenIconCaches) {
+        // The icon isn't the one we're looking for
+        if (![iconCache.iconUrl isEqualToString:request.url]) { continue; }
+        // The icon we're looking for has expired
+        if ([self.class numberOfDaysFromDateCreated:iconCache.lastModifiedDate] >= 30) { break; }
+        NSString *imageFilePath = [self.cacheFileBaseDirectory stringByAppendingPathComponent:[self.class sdl_md5HashFromString:request.url]];
+        icon = [UIImage imageWithContentsOfFile:imageFilePath];
+
+        // If we fail to get icon at path, break loop early to download new icon
+        if (icon == nil) { break; }
+        
+        return completion(icon, nil);
+    }
+
+    if (iconArchiveFile == nil || iconArchiveFile.lockScreenIconCaches.count == 0) {
```

Comment:
My understanding is that if the icon is not retrieved at the specified path, we need to clear the directory and download the icon again. I put this below the `for` because when we `break` after `icon == nil`, we will need to clear the directory and create a new archive file before downloading. Looking at it again, we should be setting `iconArchiveFile = nil` during this check. Otherwise a new archive file won't be created after the `break`.
```        
if (icon == nil) {
       SDLLogW(@"No icon was found at the path, we need to download a new icon");
       iconArchiveFile = nil;
       break;
}
```